### PR TITLE
signup page

### LINKED ISSUE #248: UI: Develop Sign Up page
## Description
As a maintainer,
I need to see BIT Sign Up page gets redesigned,
so that I can make sure they meet AnitaB guidelines.

## Mocks
Approved design idea for Sign Up page (Designer Gloria Osoro)
Screen Shot 2021-05-22 at 6 25 02 pm

### Update [Required]
- [ ] Implement the design above in the section it is intended in the design above.

### Enhancement to Update [Optional]
NA

## Definition of Done
- [ ] All of the required items are completed.
- [ ] Approval by 1 mentor.

## Estimation
1 day

**Additional Note**
Image and design metrics can be found on [Figma file](https://www.figma.com/file/V22SXBLzMd9bqnita6hfMp/Bridge_in_tech?node-id=0%3A1) under `high_fidelity` tab.

### PR DESCRIPTION
**Pull request by aakankshaagr**
### Description

Developed SignUp Page

Fixes #248 

### Type of Change:

- Code
- Quality Assurance
- User Interface

**Code/Quality Assurance Only**
- New feature (non-breaking change which adds functionality pre-approved by mentors)

### How Has This Been Tested?

image

### Checklist:

- [ ] My PR follows the style guidelines of this project
- [ ] I have performed a self-review of my own code or materials

**Code/Quality Assurance Only**
- [ ] My changes generate no new warnings
- [ ] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: src/register/Register.jsx [hunk 1]
```diff
@@ -136,71 +171,92 @@ export default function Register() {
...
-                        </form-group>
-                        <div><br></br></div>
-                        <form-group controlId="formPassword">
-                            <p className="input-control">
+                <form-group controlId="formPassword">
+                  <p className="input-control">
+                    <label id="password">Password :</label>
+                    <input
+                      aria-labelledby="password"
+                      className="feilds"
```
Comment (mtreacy002) [inline src/register/Register.jsx]: please fix typo

File: src/register/Register.jsx [hunk 2]
```diff
@@ -136,71 +171,92 @@ export default function Register() {
...
+                      type={isPasswordShown ? "text" : "password"}
+                      name="password"
+                      placeholder="atleast 8 characters"
```
Comment (mtreacy002) [inline src/register/Register.jsx]:
```suggestion
                      placeholder="at least 8 characters"
```

File: src/register/Register.jsx [hunk 3]
```diff
@@ -136,71 +171,92 @@ export default function Register() {
...
-                            {!isValidConfirmPassword && (
-                                <span className="error">Must be same as password</span>
-                            )}
-                        </form-group>
+                <form-group controlId="formPassword">
+                  <p className="input-control">
+                    <label id="confirmPassword">Confirm Password :</label>
+                    <input
+                      aria-labelledby="confirmPassword"
+                      className="feilds"
```
Comment (mtreacy002) [inline src/register/Register.jsx]: typo

File: src/register/Register.jsx [hunk 4]
```diff
@@ -136,71 +171,92 @@ export default function Register() {
...
+                      type={isPasswordShown ? "text" : "password"}
+                      name="confirmPassword"
+                      placeholder="atleast 8 characters"
```
Comment (mtreacy002) [inline src/register/Register.jsx]: although the design shows this as placehloder, IMO it makes more sense if we show `re-type password` here instead since this field is not to insert a totally new password.

File: src/register/Register.jsx [hunk 5]
```diff
@@ -2,126 +2,161 @@ import React, { useState, useContext } from "react";
...
-        setConfirmPassword(e.target.value)
-        setIsValidConfirmPassword(e.target.checkValidity() && e.target.value === password);
-    };
+  return isAuth ? (
+    <Redirect to="/" />
+  ) : (
+    <div className="container">
+      <div class="row align-items-start">
+        <div class="col">
+          <img src="src/assets/images/side_image.jpg" alt="background" />
```
Comment (mtreacy002) [inline src/register/Register.jsx]: instead of using image path and have the image saved inside images folder, you can use the githubusercontent link of an image as it was used on the [About.jsx line 12](https://github.com/anitab-org/bridge-in-tech-web/blob/develop/src/home/About.jsx).

**Review by mtreacy002 — CHANGES_REQUESTED** [review]
@aakankshaagr , thank you for submitting the PR. I have left some feedback for you below as well as an alternative to upload the side image. Also, please don't forget to remove all the commented lines that are no longer needed. Let me know if you have any questions 😉.

**Comment by aakankshaagr** [comment]
Thanks @mtreacy002 !! I have made changes as you suggested. About image I don't know how to get side_image in way suggested by you. Thankfully, somehow it worked and image is displayed. I have made changes in classes naming as there was lot of css getting included with those names.
Icon part is left as m facing some issues in icon display.
I didn't find any information for page layout when screen size is small so I made it as per my wish.If it's already there then please inform me, I will make changes according to that.
In media queries, I want screen to be divided in 2 rows top having image and bottom having form. I don't know why image is not showing up.Can you please point out where I am going wrong?

Waiting for your feedback :) !!

**Comment by aakankshaagr** [comment]
image
if Register component can too stretch like footer and navbar then it will look better.
image

**Comment by mtreacy002** [comment]
@aakankshaagr , is there any reason you don't make it stretch like the footer? I believe the image in Figma also shows that the container (image+form) is stretchable. 
Screen Shot 2021-07-08 at 11 34 09 pm

**Review by mtreacy002 — CHANGES_REQUESTED** [review]
Hi @aakankshaagr , the build compiled with warnings. Can you try to fix this, please? Thanks 
Screen Shot 2021-07-08 at 11 22 59 pm

**Comment by aakankshaagr** [comment]
@mtreacy002 On stretching image got doubled and that wasn't looking good. So, as I mentioned earlier ,I felt like only making form stretchable.

**Comment by aakankshaagr** [comment]
@mtreacy002 route.js got autoformatted and by mistake, I pushed route.js too. While trying to fix that I somehow merged develop.I don't know how to undo this. I tried to rebase and git reset --merge but it is not working. Please suggest some other way to remove this.

**Review by mtreacy002 — CHANGES_REQUESTED** [review]
@aakankshaagr , I checked your code changes and it doesn't seem to have major issues. Auto-formatting of routes.jsx seems ok and the merged of develop branch to your pr branch seems to have no impact to the code you're submitting. Although, I do have some feedback for you to work on:
* the build is failing. Below is the screenshot
Screen Shot 2021-07-16 at 1 59 39 pm

* the tests are also failing. see this [gist](https://gist.github.com/mtreacy002/9d8b61db394ff97e7c142be36f4eaa6b) for the error warnings.

Can you please fix these errors? Thanks

PS: I have a second thought on the form stretch. The stretching of the register form field seems odd since the fields on the form won't need that bigger input field. I think the form stretch on the bigger screen size should be maximized to up to 75% of the screen width instead to make it reasonable. Perhaps your screenshot [above](https://github.com/anitab-org/bridge-in-tech-web/pull/262#issuecomment-871119935) was made with smaller zoom range (~50 - 67%) which is probably not what most of our users would have (~75-125%). So, you can put it back to the way it was IMO 😉.
Screen Shot 2021-07-16 at 2 17 25 pm

**Comment by aakankshaagr** [comment]
@mtreacy002  I have removed error warnings. It is also suggesting we should add alt text for the password icon image. What shall I keep as alt text?
Form size is dynamic, as screen size increases fields will stretch. So I think changes help in case if the user screen size is way large or else it will display like previously I showed you.

**Comment by vj-codes** [comment]
@mtreacy002 hey can you please approve the workfow run in this PR?

**Comment by aakankshaagr** [comment]
Tests are failing bcz placeholder texts are not matching in tests and register files. So I am changing placeholder texts. @mtreacy002 According to your earlier suggestion password placeholder values isn't suitable for adding in tests. So I am changing them to be the same as label names. Is that fine or I should do vice versa?

**Comment by aakankshaagr** [comment]
@mtreacy002 All test cases are passed :) 
image
I hope now it passes all build tests.

**Comment by aakankshaagr** [comment]
@mtreacy002  can you please approve the workflow?

**Comment by aakankshaagr** [comment]
image

image

All test cases are passing along with no warnings. @mtreacy002 Can you plz approve the workflow.

**Comment by aakankshaagr** [comment]
@mtreacy002 can you plz approve the workflow?

**Comment by aakankshaagr** [comment]
@mtreacy002  please approve the workflow. I have changed the layout when zoom<50% as suggested by you. All test cases have been passed locally.

File: src/register/Register.jsx [hunk 6]
```diff
@@ -1,291 +1,301 @@
...
+import { BASE_API } from "../config";
+import { SERVICE_UNAVAILABLE_ERROR } from "../messages";
 import TermsAndPrivacyPolicyModal from "./TermsAndPrivacyPolicyModal";
+import hidePwd from "../assets/images/Hide_pwd_icon.svg";
+import showPwd from "../assets/images/show_pwd_icon.svg";
 
 export default function Register() {
-    const {isAuth} = useContext(AuthContext);
-    const [isValidName, setIsValidName] = useState(true);
-    const [isValidUsername, setIsValidUsername] = useState(true);
```
Comment (mtreacy002) [inline src/register/Register.jsx:12]: We need to keep this field on the form since the Mentorship System has this field in its signup form.
Screen Shot 2021-11-24 at 9 44 47 am

So in your SignUp form design, please add the username field to match the form in Mentorship System (as you can see on the existing code on BIT develop branch)

Screen Shot 2021-11-24 at 9 44 28 am
Screen Shot 2021-11-24 at 8 59 19 am

PS 1: Not having the same fields causing error (invalid payload) when user tried to register as can be seen on the screenshot above.
PS 2: Please fix the location of error response. It should be on a separate line (not on the same line as "Available as" text). You can place it on top of the SignUp button.

Also, for consistency across MS-BIT applications, you can you please follow the existing order for the fields? So start with Name, then Username, followed by Email, then continue with the rest.

**Review by mtreacy002 — CHANGES_REQUESTED** [review]
Hi @aakankshaagr , thank you for the pr and my apology for replying late. 
Please check the below feedback.
Plus, can you please keep the Sidebar and the VideoSection from the existing code in your commits. It seems like you've deleted them by accident on this pr.

Thank you beforehand 😉

**Comment by aakankshaagr** [comment]
> Hi @aakankshaagr , thank you for the pr and my apology for replying late. Please check the below feedback. Plus, can you please keep the Sidebar and the video section from the existing code in your commits. It seems like you've deleted them by accident on this pr.
> 
> Thank you beforehand 😉

It's fine @mtreacy002 😊. Sure, I will add them again. I am sorry I thought that the VideoSection part was not related to my pr and I added them by mistake. So I removed them. Thanks to other contributors for correcting my mistake.I have added username field and updated tests cases too. Also repositioned error before signup button. Can you please check the changes once again?
Also, I found that footer is not working. Link tag is missing `to` attribute and this is causing a warning in the register form section. Please look into it once and tell me if there is something I can do about it.

**Comment by aakankshaagr** [comment]
image
@mtreacy002 is this matching with your suggested changes? Please review this and let me know if there are some changes that need to be done.
Thank you

File: src/register/Register.jsx [hunk 7]
```diff
@@ -1,291 +1,331 @@
...
-    const [isValidPassword, setIsValidPassword] = useState(true);
-    const [isValidConfirmPassword, setIsValidConfirmPassword] = useState(true);
-    const [responseMessage, setResponseMessage] = useState(null);
-    const [password, setPassword] = useState("");
-    const [confirmPassword, setConfirmPassword] = useState("");
-    const [isPasswordShown, setIsPasswordShown] = useState(false);
-    const [showTermsModal,setShowTermsModal] = useState(false);
+  const { isAuth } = useContext(AuthContext);
+  const [isValidUsername, setIsValidUsername] = useState(true);
+  const [isValidName, setIsValidName] = useState(true);
```
Comment (rkraeher) [inline src/register/Register.jsx:15]: if these are validators, does it make more sense for initial state to be set to `false` instead? 🤔  maybe they should only be `true` after successful validation?

File: src/register/Register.jsx [hunk 8]
```diff
@@ -1,291 +1,331 @@
...
+    <React.Fragment>
+      <section className="screen">
+        <section className="split-screen">
+          <div className="left"></div>
+          <div className="right">
+            <form id="form" onSubmit={handleSubmit}>
+              <section className="register">
+                <h1>Welcome To Bridge In Tech</h1>
+                <h4>
+                  Lets start your account setup so you can have access to a
```
Comment (rkraeher) [inline src/register/Register.jsx]: Lets is a typo. should be Let's

File: src/tests/PersonalDetails.test.js [hunk 9]
```diff
@@ -1,144 +1,212 @@
...
-    expect(screen.getByLabelText("Interests :", {selector: "input"})).toHaveValue("coding");
-    expect(screen.getByLabelText("Bio :", {selector: "input"})).toHaveValue("Student");
-    expect(screen.getByLabelText("Occupation :", {selector: "input"})).toHaveValue("Student");
-    expect(screen.getByLabelText("Location :", {selector: "input"})).toHaveValue("India");
-    expect(screen.getByLabelText("Current Organization :", {selector: "input"})).toHaveValue("University");
-    expect(screen.getByLabelText("Skills :", {selector: "input"})).toHaveValue("Tech, Open Source");
-    expect(screen.getByLabelText("Photo Url :", {selector: "input"})).toBeEmpty();
-    expect(screen.getByLabelText("Resume Url :", {selector: "input"})).toBeEmpty();
+    expect(
+      screen.getByLabelText("Username :", { selector: "input" })
```
Comment (rkraeher) [inline src/tests/PersonalDetails.test.js:57]: there is a lot of weird formatting and line breaks in this file that really decreases the readability of the tests.

File: src/tests/Register.test.js [hunk 10]
```diff
@@ -1,202 +1,321 @@
...
-import { rest } from 'msw';
-import { setupServer } from 'msw/node';
-import { render, fireEvent, screen, wait, waitForElement } from '@testing-library/react';
-import '@testing-library/jest-dom';
-import { act } from 'react-dom/test-utils';
+import React from "react";
+import { rest } from "msw";
+import { setupServer } from "msw/node";
+import {
+  render,
```
Comment (rkraeher) [inline src/tests/Register.test.js:5]: something odd happened with the formatting in this file as well. this negatively affects readability like ☝️

File: src/tests/Register.test.js [hunk 11]
```diff
@@ -1,202 +1,321 @@
...
+    <Router>
+      <Register />
+    </Router>
+  );
+
+  fireEvent.change(screen.getByPlaceholderText("Full Name"), {
+    target: { value: "My Name" },
+  });
+  act(() => {
+    fireEvent.change(screen.getByPlaceholderText("Username"), {
```
Comment (rkraeher) [inline src/tests/Register.test.js:53]: do we really need to use `act()` in this test if none of these form input events are asynchronous except for the submit registration?

**Review by rkraeher — COMMENTED** [review]
nice work 🙂 just a refactor suggestion: there are a lot of repeated assertions throughout the tests where the only thing that changes are the values you are passing. you could remove a lot of duplicated code by mapping these repeated values with the jest `it.each` method. [see the docs](https://jestjs.io/docs/api#testeachtablename-fn-timeout)

File: src/register/Register.jsx (hunk 7 above)
Comment (aakankshaagr) [reply src/register/Register.jsx:15]: @rkraeher Thank you for reviewing. Values are set so when the user enters the wrong value then display an error. Apart from validating, error message is also handled by this.

File: src/tests/Register.test.js (hunk 11 above)
Comment (aakankshaagr) [reply src/tests/Register.test.js:53]: @rkraeher I guess we need to use act() 
image

File: src/tests/Register.test.js (hunk 10 above)
Comment (aakankshaagr) [reply src/tests/Register.test.js:5]: I guess it's because prettier is not working properly for me. I guess the code will be readable after running the prettier. I tried hard to correct this even done manually but after saving everything goes back to the same.So it's quite difficult to format tests.

Comment (rkraeher) [reply src/tests/Register.test.js:5]: I see. do you have prettier extension or any other linter enabled? it is annoying to do it manually but if you do that and have the prettier disabled it shouldn't autoformat when you save

File: src/tests/Register.test.js (hunk 11 above)
Comment (rkraeher) [reply src/tests/Register.test.js:53]: this warning seems to have to do with state updates happening asynchronously. this is not really how `act` should be used if you check out the [docs](https://reactjs.org/docs/test-utils.html#act). 

[This article](https://davidwcai.medium.com/react-testing-library-and-the-not-wrapped-in-act-errors-491a5629193b) also has a really useful explanation of the error and some suggestions on alternative solutions for how to deal with it effectively and eliminate the warning
